File: backend/app/services/supabase_service.py
from supabase import create_client
import os
from typing import List, Dict, Any

class SupabaseService:
    client = None

    @classmethod
    def get_client(cls):
        if cls.client is None:
            url = os.getenv('SUPABASE_URL')
            key = os.getenv('SUPABASE_ANON_KEY')
            cls.client = create_client(url, key)
        return cls.client

    @staticmethod
    def get_groups() -> List[Dict[str, Any]]:
        client = SupabaseService.get_client()
        response = client.table('groups').select('*').execute()
        return response.data

    @staticmethod
    def share_to_group(group_id: str, news_id: str, user_id: str = "anonymous") -> Dict[str, Any]:
        """Share news to a group by creating a message"""
        client = SupabaseService.get_client()
        message_data = {
            'group_id': group_id,
            'user_id': user_id,
            'content': f'Shared news article',
            'news_id': news_id
        }
        response = client.table('messages').insert(message_data).execute()
        return response.data

    @staticmethod
    def create_group(name: str, description: str, group_type: str) -> Dict[str, Any]:
        client = SupabaseService.get_client()
        group_data = {'name': name, 'description': description, 'type': group_type}
        response = client.table('groups').insert(group_data).execute()
        return response.data
    
    @staticmethod
    def get_group_messages(group_id: str, limit: int = 50) -> List[Dict[str, Any]]:
        """Get messages for a specific group"""
        client = SupabaseService.get_client()
        response = client.table('messages').select('*').eq('group_id', group_id).order('created_at', desc=True).limit(limit).execute()
        return response.data
    
    @staticmethod
    def send_message(group_id: str, user_id: str, content: str, news_id: str = None) -> Dict[str, Any]:
        """Send a message to a group"""
        client = SupabaseService.get_client()
        message_data = {
            'group_id': group_id,
            'user_id': user_id,
            'content': content
        }
        if news_id:
            message_data['news_id'] = news_id
        response = client.table('messages').insert(message_data).execute()
        return response.data

File: backend/services/supabaseService.js
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';

dotenv.config();

const supabase = createClient(process.env.SUPABASE_URL, process.env.SUPABASE_ANON_KEY);

export const saveNewsToSupabase = async (articles) => {
  try {
    const { data, error } = await supabase
      .from('news_articles')
      .upsert(articles, { onConflict: 'id', ignoreDuplicates: false });
    
    if (error) throw error;
    
    console.log(`Saved ${articles.length} articles to Supabase.`);
    return data;
  } catch (error) {
    console.error('Error saving news:', error);
    return null;
  }
};

export const getNewsFromSupabase = async (sector = null, limit = 50) => {
  try {
    let query = supabase
      .from('news_articles')
      .select('*')
      .order('published_at', { ascending: false })
      .limit(limit);
    
    if (sector && sector !== 'General') {
      query = query.eq('sector', sector);
    }
    
    const { data, error } = await query;
    
    if (error) throw error;
    
    console.log(`Fetched ${data?.length || 0} articles from Supabase`);
    return data || [];
  } catch (error) {
    console.error('Error fetching news:', error);
    return [];
  }
};

export const searchNewsInSupabase = async (queryText) => {
  try {
    const { data, error } = await supabase
      .from('news_articles')
      .select('*')
      .or(`title.ilike.%${queryText}%,summary.ilike.%${queryText}%,content.ilike.%${queryText}%`)
      .limit(20);
    
    if (error) throw error;
    
    return data || [];
  } catch (error) {
    console.error('Error searching news:', error);
    return [];
  }
};

export const getGroups = async () => {
  const { data, error } = await supabase.from('groups').select('*');
  if (error) throw error;
  return data || [];
};

export const createGroup = async (groupData) => {
  const { data, error } = await supabase.from('groups').insert([groupData]).select();
  if (error) throw error;
  return data;
};

export const getGroupMessages = async (groupId, limit = 50) => {
  const { data, error } = await supabase
    .from('group_messages')
    .select('*, news_articles(*)')
    .eq('group_id', groupId)
    .order('created_at', { ascending: true })
    .limit(limit);
  
  if (error) throw error;
  return data || [];
};

export const sendMessage = async (messageData) => {
  const { data, error } = await supabase.from('group_messages').insert([messageData]).select();
  if (error) throw error;
  return data;
};

export const shareArticleToGroup = async (groupId, newsId, userId) => {
  const { data, error } = await supabase.from('group_messages').insert([{
    group_id: groupId,
    user_id: userId,
    content: "Shared an article",
    news_id: newsId
  }]).select();
  
  if (error) throw error;
  return data;
};
